packages/grafana-data/src/utils/vector.ts

```diff
@@ -109,18 +99,106 @@ export class CircularVector<T = any> implements Vector<T> {
   }
 }
 
+interface CircularOptions<T> {
+  buffer?: T[];
+  append?: 'head' | 'tail';
+  capacity?: number;
+}
+
 /**
- * Values are returned in the order defined by the input parameter
+ * Circular vector uses a single buffer to capture a stream of values
+ * overwriting the oldest value on add.
+ *
+ * This supports addting to the 'head' or 'tail' and will grow the buffer
+ * to match a configured capacity.
  */
-export class SortedVector<T = any> implements Vector<T> {
-  constructor(private source: Vector<T>, private order: number[]) {}
+export class CircularVector<T = any> implements Vector<T> {
+  private buffer: T[];
+  private index: number;
+  private capacity: number;
+  private tail: boolean;
+
+  constructor(options: CircularOptions<T>) {
+    this.buffer = options.buffer || [];
+    this.capacity = this.buffer.length;
+    this.tail = 'head' !== options.append;
+    this.index = 0;
 
-  get length(): number {
-    return this.source.length;
+    this.add = this.getAddFunction();
+    if (options.capacity) {
+      this.setCapacity(options.capacity);
+    }
   }
 
-  get(index: number): T {
-    return this.source.get(this.order[index]);
+  // Pick an add function that either writes forwards or backwards
+  private getAddFunction() {
+    if (this.tail) {
+      return (value: T) => {
+        this.buffer[this.index] = value;
```

Comment:
Do we need the push/unshift then change to this? I mean seems like this two would actually work from the start. If it is a perf optimisation (so the initial fill is faster), is it significant one?

Reply:
I mean we could remove the swapping of implementation based on capacity and just have `getAddFunction` content's directly in the `add` function. Like:
```
add () {
  if (this.tail) {
    this.buffer[this.index] = value;
    this.index = (this.index + 1) % this.buffer.length;
  } else {
    let idx = this.index - 1;
    if (idx < 0) {
        idx = this.buffer.length - 1;
      }
      this.buffer[idx] = value;
      this.index = idx;
  }
}
```
Or if we want to get rid of the `if` on each call we can assign the proper function at construction.